azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/sqlserver/AlwaysEncryptedAutoConfiguration.java

```diff
@@ -0,0 +1,109 @@
+import org.springframework.beans.FatalBeanException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties;
+import org.springframework.boot.autoconfigure.jdbc.JndiDataSourceAutoConfiguration;
+import org.springframework.boot.autoconfigure.jdbc.XADataSourceAutoConfiguration;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
+import org.springframework.util.ClassUtils;
+
+import javax.sql.DataSource;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+@ConditionalOnClass({DataSource.class, EmbeddedDatabaseType.class})
+@EnableConfigurationProperties({DataSourceProperties.class, KeyVaultProperties.class})
+@ConditionalOnProperty(name = AEConstants.PROPERTY_AE_ENABLED, matchIfMissing = false)
+public class AlwaysEncryptedAutoConfiguration {
+    private static final Logger LOG = LoggerFactory.getLogger(AlwaysEncryptedAutoConfiguration.class);
+
+    private final KeyVaultProperties properties;
+    private final TelemetryProxy telemetryProxy;
+
+    public AlwaysEncryptedAutoConfiguration(KeyVaultProperties properties) {
+        this.properties = properties;
+        this.telemetryProxy = new TelemetryProxy(properties.isAllowTelemetry());
+
+    }
+
+    private void trackCustomEvent() {
+        final HashMap<String, String> customTelemetryProperties = new HashMap<>();
+
+        final String[] packageNames = this.getClass().getPackage().getName().split("\\.");
+
+        if (packageNames.length > 1) {
+            customTelemetryProperties.put(TelemetryData.SERVICE_NAME, packageNames[packageNames.length - 1]);
+        }
+        telemetryProxy.trackEvent(ClassUtils.getUserClass(this.getClass()).getSimpleName(), customTelemetryProperties);
+    }
+
+    @Bean(name = "dataSourceBeanPostProcessor")
+    @ConditionalOnClass(com.microsoft.sqlserver.jdbc.SQLServerDriver.class)
+    @ConditionalOnBean(DataSource.class)
+    public BeanPostProcessor dataSourceBeanPostProcessor() {
+        return new BeanPostProcessor() {
```

Comment:
This bean post processor is better to be defined as separate class.

Reply:
I'm unsure whether this will work. My understanding is that `BeanPostProcessor` should be initialized in the very early stage, otherwise it won't take effect.